File: tests/test_7_4_2_string_steps.py
"""
Tests for Task 7.4.2: String Manipulation Traversal Steps.

All tests call string steps (concat, toLower, toUpper, trim, replace,
split, length, substring, reverse, format) on a live TinkerCat traversal.
None of these steps are implemented yet; every test fails with
AttributeError until they are added to TinkerCat's GraphTraversal.
See docs/project/changelog/task-7.4.2-string-steps.adoc.
"""

import pytest

from tinkercat import TinkerCat


@pytest.fixture
def g():
    graph = TinkerCat()
    graph.add_vertex("word", name="Hello")
    graph.add_vertex("word", name="HELLO")
    graph.add_vertex("word", name="hello")
    graph.add_vertex("word", name="  hello  ")
    graph.add_vertex("word", name="foo bar")
    graph.add_vertex("word", name="a,b,c")
    graph.add_vertex("word", name="a, b, c")
    graph.add_vertex("word", name="HÉLLO")
    graph.add_vertex("word", name="  alice@EXAMPLE.COM  ")
    yield graph
    graph.close()


def test_concat(g):
    result = g.traversal().V().has("name", "Hello").values("name").concat(", World").to_list()  # AttributeError
    assert result == ["Hello, World"]


def test_concat_multiple(g):
    result = g.traversal().V().has("name", "Hello").values("name").concat("b", "c").to_list()  # AttributeError
    assert result == ["Hellobc"]


def test_format_percent_s(g):
    result = g.traversal().V().has("name", "Hello").values("name").format("Hi, %s!").to_list()  # AttributeError
    assert result == ["Hi, Hello!"]


def test_format_tokens(g):
    result = g.traversal().V().has("name", "Hello").values("name").format("%{val}").to_list()  # AttributeError
    assert isinstance(result, list)


def test_format_missing_token_empty(g):
    result = g.traversal().V().has("name", "Hello").values("name").format("%{missing}").to_list()  # AttributeError
    assert result == [""]


def test_to_lower(g):
    result = g.traversal().V().has("name", "HELLO").values("name").to_lower().to_list()  # AttributeError
    assert result == ["hello"]


def test_to_upper(g):
    result = g.traversal().V().has("name", "hello").values("name").to_upper().to_list()  # AttributeError
    assert result == ["HELLO"]


def test_trim(g):
    result = g.traversal().V().has("name", "  hello  ").values("name").trim().to_list()  # AttributeError
    assert result == ["hello"]


def test_ltrim(g):
    result = g.traversal().V().has("name", "  hello  ").values("name").ltrim().to_list()  # AttributeError
    assert result == ["hello  "]


def test_rtrim(g):
    result = g.traversal().V().has("name", "  hello  ").values("name").rtrim().to_list()  # AttributeError
    assert result == ["  hello"]


def test_replace(g):
    result = g.traversal().V().has("name", "foo bar").values("name").replace("bar", "baz").to_list()  # AttributeError
    assert result == ["foo baz"]


def test_split_delimiter(g):
    result = g.traversal().V().has("name", "a,b,c").values("name").split(",").to_list()  # AttributeError
    assert result == ["a", "b", "c"]


def test_split_then_trim(g):
    result = g.traversal().V().has("name", "a, b, c").values("name").split(",").trim().to_list()  # AttributeError
    assert result == ["a", "b", "c"]


def test_length(g):
    result = g.traversal().V().has("name", "hello").values("name").length().to_list()  # AttributeError
    assert result == [5]


def test_length_empty_string():
    g = TinkerCat()
    try:
        g.add_vertex("w", name="")
        result = g.traversal().V().values("name").length().to_list()  # AttributeError
        assert result == [0]
    finally:
        g.close()


def test_substring_start_only(g):
    result = g.traversal().V().has("name", "hello").values("name").substring(2).to_list()  # AttributeError
    assert result == ["llo"]


def test_substring_start_end(g):
    result = g.traversal().V().has("name", "hello").values("name").substring(1, 4).to_list()  # AttributeError
    assert result == ["ell"]


def test_reverse(g):
    result = g.traversal().V().has("name", "hello").values("name").reverse().to_list()  # AttributeError
    assert result == ["olleh"]


def test_reverse_empty():
    g = TinkerCat()
    try:
        g.add_vertex("w", name="")
        result = g.traversal().V().values("name").reverse().to_list()  # AttributeError
        assert result == [""]
    finally:
        g.close()


def test_unicode_to_lower(g):
    result = g.traversal().V().has("name", "HÉLLO").values("name").to_lower().to_list()  # AttributeError
    assert result == ["héllo"]


def test_pipeline_composition(g):
    result = (
        g.traversal().V()
        .has("name", "  alice@EXAMPLE.COM  ")
        .values("name")
        .trim()
        .to_lower()
        .to_list()  # AttributeError
    )
    assert result == ["alice@example.com"]

File: tests/kotlin/test_7_4_2_string_steps.kt
package org.apache.tinkerpop.gremlin.tinkercat.tests

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import org.apache.tinkerpop.gremlin.tinkercat.structure.TinkerCat
import org.apache.tinkerpop.gremlin.tinkercat.structure.TinkerVertex

/**
 * Tests for Task 7.4.2: String Manipulation Traversal Steps.
 *
 * Every test calls a string step (concat, format, toLower, toUpper, trim, ltrim, rtrim,
 * replace, split, length, substring, reverse) on a live TinkerCat traversal.
 * Each step throws UnsupportedOperationException until implemented.
 * Mirrors tests/python/test_7_4_2_string_steps.py.
 */
class Test_7_4_2_StringSteps : StringSpec({

    lateinit var graph: TinkerCat

    beforeTest {
        graph = TinkerCat.open()
        for (name in listOf("Hello", "HELLO", "hello", "  hello  ", "foo bar",
                             "a,b,c", "a, b, c", "HÉLLO", "  alice@EXAMPLE.COM  ")) {
            graph.addVertex(TinkerVertex.T.label, "word", "name", name)
        }
    }

    "concat appends a single suffix" {
        val result = graph.traversal().V()
            .has("name", "Hello").values<String>("name")
            .concat(", World")
            .toList()
        result shouldBe listOf("Hello, World")
    }

    "concat appends multiple parts" {
        val result = graph.traversal().V()
            .has("name", "Hello").values<String>("name")
            .concat("b", "c")
            .toList()
        result shouldBe listOf("Hellobc")
    }

    "format with percent-s placeholder" {
        val result = graph.traversal().V()
            .has("name", "Hello").values<String>("name")
            .format("Hi, %s!")
            .toList()
        result shouldBe listOf("Hi, Hello!")
    }

    "format with token placeholder" {
        val result = graph.traversal().V()
            .has("name", "Hello").values<String>("name")
            .format("%{val}")
            .toList()
        (result is List<*>) shouldBe true
    }

    "format with missing token yields empty string" {
        val result = graph.traversal().V()
            .has("name", "Hello").values<String>("name")
            .format("%{missing}")
            .toList()
        result shouldBe listOf("")
    }

    "toLower converts HELLO to hello" {
        val result = graph.traversal().V()
            .has("name", "HELLO").values<String>("name")
            .toLower()
            .toList()
        result shouldBe listOf("hello")
    }

    "toUpper converts hello to HELLO" {
        val result = graph.traversal().V()
            .has("name", "hello").values<String>("name")
            .toUpper()
            .toList()
        result shouldBe listOf("HELLO")
    }

    "trim strips both ends" {
        val result = graph.traversal().V()
            .has("name", "  hello  ").values<String>("name")
            .trim()
            .toList()
        result shouldBe listOf("hello")
    }

    "ltrim strips leading whitespace only" {
        val result = graph.traversal().V()
            .has("name", "  hello  ").values<String>("name")
            .ltrim()
            .toList()
        result shouldBe listOf("hello  ")
    }

    "rtrim strips trailing whitespace only" {
        val result = graph.traversal().V()
            .has("name", "  hello  ").values<String>("name")
            .rtrim()
            .toList()
        result shouldBe listOf("  hello")
    }

    "replace substitutes a substring" {
        val result = graph.traversal().V()
            .has("name", "foo bar").values<String>("name")
            .replace("bar", "baz")
            .toList()
        result shouldBe listOf("foo baz")
    }

    "split on delimiter emits tokens" {
        val result = graph.traversal().V()
            .has("name", "a,b,c").values<String>("name")
            .split(",")
            .toList()
        result shouldBe listOf("a", "b", "c")
    }

    "split then trim removes extra whitespace" {
        val result = graph.traversal().V()
            .has("name", "a, b, c").values<String>("name")
            .split(",")
            .trim()
            .toList()
        result shouldBe listOf("a", "b", "c")
    }

    "length returns character count" {
        val result = graph.traversal().V()
            .has("name", "hello").values<String>("name")
            .length()
            .toList()
        result shouldBe listOf(5)
    }

    "length of empty string is zero" {
        val g2 = TinkerCat.open()
        try {
            g2.addVertex(TinkerVertex.T.label, "w", "name", "")
            val result = g2.traversal().V().values<String>("name").length().toList()
            result shouldBe listOf(0)
        } finally {
            g2.close()
        }
    }

    "substring with start only" {
        val result = graph.traversal().V()
            .has("name", "hello").values<String>("name")
            .substring(2)
            .toList()
        result shouldBe listOf("llo")
    }

    "substring with start and end" {
        val result = graph.traversal().V()
            .has("name", "hello").values<String>("name")
            .substring(1, 4)
            .toList()
        result shouldBe listOf("ell")
    }

    "reverse flips a string" {
        val result = graph.traversal().V()
            .has("name", "hello").values<String>("name")
            .reverse()
            .toList()
        result shouldBe listOf("olleh")
    }

    "reverse of empty string is empty string" {
        val g2 = TinkerCat.open()
        try {
            g2.addVertex(TinkerVertex.T.label, "w", "name", "")
            val result = g2.traversal().V().values<String>("name").reverse().toList()
            result shouldBe listOf("")
        } finally {
            g2.close()
        }
    }

    "toLower handles unicode" {
        val result = graph.traversal().V()
            .has("name", "HÉLLO").values<String>("name")
            .toLower()
            .toList()
        result shouldBe listOf("héllo")
    }

    "pipeline: trim then toLower" {
        val result = graph.traversal().V()
            .has("name", "  alice@EXAMPLE.COM  ").values<String>("name")
            .trim()
            .toLower()
            .toList()
        result shouldBe listOf("alice@example.com")
    }
})
